subprojects/platform-jvm/src/main/java/org/gradle/jvm/toolchain/JavaLanguageVersion.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright 2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.jvm.toolchain;
+
+import org.gradle.api.Incubating;
+import org.gradle.jvm.toolchain.internal.DefaultJavaLanguageVersion;
+
+/**
+ * Represents a Java Language version
+ *
+ * @since 6.7
+ */
+@Incubating
+public interface JavaLanguageVersion extends Comparable<JavaLanguageVersion> {
+
+    static JavaLanguageVersion of(int version) {
+        return new DefaultJavaLanguageVersion(version);
+    }
+
+    static JavaLanguageVersion of(String version) {
+        return new DefaultJavaLanguageVersion(Integer.parseInt(version));
+    }
+
+    /**
+     * Return this version as a number, 14 for Java 14.
+     *
+     * @return the version number
+     */
+    int asInt();
+
+    /**
+     * Return this version as a String, "14" for Java 14.
+     *
+     * @return the version number
+     */
+    String asString();
+
+    /**
+     * Indicates if this version can interoperate with the other version.
+     * <p>
+     * For example, Java 14 interoperates with Java 11, but not the opposite.
+     *
+     * @param other the language version to interoperate with
+     *
+     * @return {@code true} if this version can interoperate with the other version, {@code false} otherwise
+     */
+    boolean interoperatesWith(JavaLanguageVersion other);
```

Comment:
:o: `isBackwardsCompatibleWith`?

Reply:
Thanks for the suggestion. I do not like the `Backwards` too much. I also no longer like what I chose because `interoperate` contains an element of symmetry, which is definitely not the case here.
@bmuskalla also indicated that the meaning of the similar method in `JavaVersion` was not always clear.
It is however quite clear in my mind: Can _this_ language version understand source or bytecode in _that_ language version. It is definitely not about production.
Maybe something like: `canInterpret` which can be applied to source and bytecode.

Reply:
`canInterpret` is still quite obscure to me. I prefer `compatibleWith` TBH. We could use things like `canCompileAgainst` or `canUnderstandClassesCreatedWith`

Reply:
`canCompileOrRun` then to be fully explicit about it?

Reply:
As mentioned in the call, I think this method is hard to name right because it heavily depends on the context that "compatible" means. Does it mean "Version X can run classes from Y" or "Version X can use release flag for Y" or "Version X has equal language features as Y"). Every time I used the same APIs on `JavaVersion`, after reading the javadocs, I ended up inlining the `X <= Y` into the call site as it was more readable what the intention was.

Reply:
That's why I want a name that makes it crystal clear that the context is what I said above:

> Can this language version understand source or bytecode in that language version?